feat: Compress private and public files during backup

gavindsouza opened the pull request:
### Updates

* Pass `--compress` to the backup command to create tar archive files.
**eg:** bench --site gavin.frappe.cloud backup --with-files --compress

* Added missing `--backup-path-db`, `--backup-path-conf`, `--backup-path-files`, `--backup-path-private-files` in the `backup` command.

* Added a new `--backup-path` to allow setting folder for where all backups can be saved.

* `--backup-path-db`, `--backup-path-conf`, `--backup-path-files`, `--backup-path-private-files` have higher priority with respect to `--backup-path`. Backups will be saved under the file name explicitly stated using the `--backup-path-*` options with files saved under the folder path dictated by `--backup-path` or the default folder path (`./sites/{name}/private/backups`) as fallback.

* Added success/failure messages irrespective of verbosity
* Always show file details in the backup summary after successful operations. 
* Use the `verbose` flag to show internal functions [WIP across frappe CLI]

----

### Usage

![Screenshot 2020-09-01 at 8 46 47 PM](https://user-images.githubusercontent.com/36654812/91870150-509cf500-ec94-11ea-842e-3490694ee96f.png)

----

### Changes

- `--compress` flag

Before:
![Screenshot 2020-09-01 at 4 06 48 PM](https://user-images.githubusercontent.com/36654812/91839747-34399200-ec6d-11ea-9437-f715b8727b20.png)

After:
![Screenshot 2020-09-01 at 4 06 58 PM](https://user-images.githubusercontent.com/36654812/91839744-326fce80-ec6d-11ea-8219-787a6b10c28f.png)

- Verbosity updates

![Screenshot 2020-09-28 at 7 08 17 PM](https://user-images.githubusercontent.com/36654812/94439399-fe30f480-01bd-11eb-8238-74d7383a778d.png)

![Screenshot 2020-09-28 at 7 08 09 PM](https://user-images.githubusercontent.com/36654812/94439406-fffab800-01bd-11eb-9e3d-ee3d7d31dc6a.png)


---

**Note:** Compressed files aren't accessible via the `BackupGenerator.get_recent_backup` method or `frappe.utils.backups.fetch_latest_backups` API yet.

Docs Link: https://github.com/frappe/frappe_docs/pull/48

**ps:** there were empty files that I'd created of a couple of Gs. Hence the dramatic size difference.

gavindsouza committed:
feat: Compress private and public files during backup

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into compress-files

gavindsouza committed:
fix: Sider

gavindsouza committed:
fix: Pass conf path via backup command

gavindsouza committed:
fix: Add command options and description

netchampfaris commented on frappe/commands/site.py:
```diff
@@ -388,10 +388,15 @@ def use(site, sites_path='.'):
 
 @click.command('backup')
 @click.option('--with-files', default=False, is_flag=True, help="Take backup with files")
+@click.option('--backup-path-db', default=None, help="Compress private and public files")
```
Would like to see `--backup-path` option that would store all backup files here

gavindsouza replied:
Sounds cool. Added!

gavindsouza committed:
fix: Set correct help for options

gavindsouza committed:
feat: Allow setting folder for where all backups can be set

Passing --backup-path {folder_name} to backup will set where the backup
files will be saved

gavindsouza committed:
chore: Optimize imports

gavindsouza committed:
fix: Added base verbosity for success/failure in backup command

gavindsouza committed:
fix: Show failure message at all costs

gavindsouza committed:
fix: Make dirs if dont exist

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into compress-files

gavindsouza committed:
Merge branch 'develop' into compress-files

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into compress-files

gavindsouza committed:
fix: Folder creation failures, click.(s)echo

gavindsouza committed:
test: Added tests for backup command

gavindsouza committed:
fix: Enhancements for the BaseCommands class

gavindsouza committed:
test: Fix tests to map with functionality

netchampfaris commented:
@gavindsouza I tried the command.

<img width="723" alt="image" src="https://user-images.githubusercontent.com/9355208/94431308-ba84bd80-01b2-11eb-99a7-316a11f7776f.png">

I guess, it worked. The output is confusing.

Can we also print the file paths with their file sizes that were created with the command?

gavindsouza commented:
@netchampfaris That output is from the `-v` I'd added while passing the files in gzip. I've removed that now.

Also, now there's always gonna be a summary printed irrespective of passing `--verbose`.

![Screenshot 2020-09-28 at 7 08 17 PM](https://user-images.githubusercontent.com/36654812/94439399-fe30f480-01bd-11eb-8238-74d7383a778d.png)
![Screenshot 2020-09-28 at 7 08 09 PM](https://user-images.githubusercontent.com/36654812/94439406-fffab800-01bd-11eb-9e3d-ee3d7d31dc6a.png)

How's this?

gavindsouza committed:
fix: Verbosity. Show Backup summary, always

gavindsouza committed:
fix: Added human readable file size func

gavindsouza committed:
chore: Optimized imports

gavindsouza committed:
fix: Sider

netchampfaris commented:
@gavindsouza very cool!

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into compress-files

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into compress-files
